File: eval/fixtures/ruby_demo/user_repository.rb
class UserRepository
  # CWE-89: a raw SQL fragment built by string interpolation straight
  # from `params` is handed to ActiveRecord's `where`, which rules/vendor/
  # semgrep-rules/ruby/rails/security/injection/tainted-sql-string.yaml
  # matches directly (mode: taint, source `params`/`request`). Unlike
  # CommandService's own sink, this rule's source is `params` itself, not
  # "any parameter" -- so the reference has to be written here rather than
  # passed in as an argument, even though this is a plain service object,
  # not a controller, at runtime.
  def find_user
    User.where("username = '#{params[:username]}'")
  end

  def find_user_safe
    User.where(username: params[:username])
  end
end

File: eval/fixtures/php_demo/UserRepository.php
<?php

namespace App\Services;

use Doctrine\DBAL\Connection;

class UserRepository
{
    public static function findUser(Connection $connection, $username)
    {
        // CWE-89: the query is built by string concatenation from a
        // username argument, straight into Connection::executeQuery() --
        // the shape rules/vendor/semgrep-rules/php/doctrine/security/audit/
        // doctrine-dbal-dangerous-query.yaml matches directly (a plain
        // pattern rule keyed on the Doctrine\DBAL\Connection import above,
        // not mode: taint).
        $sql = "SELECT * FROM users WHERE username = '" . $username . "'";
        return $connection->executeQuery($sql);
    }

    public static function findUserSafe(Connection $connection, $username)
    {
        return $connection->executeQuery("SELECT * FROM users WHERE username = ?", [$username]);
    }
}
